apphub/instance_detection/yolov5/yolov5_tf.py

```diff
@@ -0,0 +1,579 @@
+# Copyright 2021 The FastEstimator Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import math
+import random
+import tempfile
+
+import cv2
+import numpy as np
+import tensorflow as tf
+from albumentations import BboxParams
+from tensorflow.keras import layers
+from tensorflow.keras.initializers import Constant
+from torch.utils.data import Dataset
+
+import fastestimator as fe
+from fastestimator.dataset.data import mscoco
+from fastestimator.op.numpyop import Delete, NumpyOp
+from fastestimator.op.numpyop.meta import Sometimes
+from fastestimator.op.numpyop.multivariate import CenterCrop, HorizontalFlip, LongestMaxSize, PadIfNeeded
+from fastestimator.op.numpyop.univariate import ReadImage, ToArray
+from fastestimator.op.tensorop import Average, TensorOp
+from fastestimator.op.tensorop.model import ModelOp, UpdateOp
+from fastestimator.schedule import EpochScheduler, cosine_decay
+from fastestimator.trace.adapt import LRScheduler
+from fastestimator.trace.io import BestModelSaver
+from fastestimator.trace.metric import MeanAveragePrecision
+from fastestimator.util import get_num_devices
+
+
+# This dataset selects 4 images and its bboxes
+class PreMosaicDataset(Dataset):
+    def __init__(self, mscoco_ds):
+        self.mscoco_ds = mscoco_ds
+
+    def __len__(self):
+        return len(self.mscoco_ds)
+
+    def __getitem__(self, idx):
+        indices = [idx] + [random.randint(0, len(self) - 1) for _ in range(3)]
```

Comment:
does it matter that this could draw the same image multiple times? Or is probability low enough to just not bother avoiding?

Reply:
the probability is 1/120k, but it's ok to draw the same image.